scanner.go

```diff
@@ -148,32 +149,34 @@ var structInfoCache = &sync.Map{}
 
 func getStructInfo(targetStruct interface{}) (reflect.Type, reflect.Value, []Field, error) {
 	v := reflect.ValueOf(targetStruct)
-	t := v.Type()
 
-	if t.Kind() != reflect.Ptr {
-		return nil, reflect.Value{}, nil, fmt.Errorf("expected struct pointer but got: %T", targetStruct)
+	t, fields, err := getStructInfoForType(v.Type())
+	if err != nil {
+		return nil, reflect.Value{}, nil, err
 	}
+
+	// Only validate v after parsing the type, otherwise the call to v.IsNil() might panic.
 	if v.IsNil() {
 		return nil, reflect.Value{}, nil, fmt.Errorf("expected non-nil pointer to struct, but got: %#v", targetStruct)
 	}
 
-	t = t.Elem()
-	if t.Kind() != reflect.Struct {
-		return nil, reflect.Value{}, nil, fmt.Errorf("can only get struct info from structs, but got: %#v", targetStruct)
-	}
-	fields, err := getStructInfoForType(t)
 	return t, v, fields, err
 }
 
-func getStructInfoForType(t reflect.Type) ([]Field, error) {
-	if t.Kind() != reflect.Struct {
-		return nil, fmt.Errorf("can only get struct info from structs, but got: %#v", t.String())
-	}
-
-	data, _ := structInfoCache.Load(t)
+func getStructInfoForType(ptrType reflect.Type) (reflect.Type, []Field, error) {
+	data, _ := structInfoCache.Load(ptrType)
 	info, ok := data.([]Field)
 	if ok {
-		return info, nil
+		return ptrType.Elem(), info, nil
```

Comment:
A possible tiny optimization as we only store a single type in the Cache:

```suggestion
	data, ok := structInfoCache.Load(ptrType)
	if ok {
		return ptrType.Elem(), data.([]Field), nil
	}
```

Might not be worth it as this could panic at runtime if we get it wrong.

Reply:
I think it is pretty safe, let me do it.

In general I think a panic is only acceptable if it is caused by a code error, which matches this situation well.

Reply:
Done